### PR TITLE
Add in early exit to PACG. Make RCG be state machine

### PR DESCRIPTION
**Pull request by tgerdesnv**
(manually cherry-picked in after review and squash)

### CODE REVIEW

File: model_analyzer/config/generate/perf_analyzer_config_generator.py [hunk 1]
```diff
@@ -15,12 +15,18 @@
...
+from model_analyzer.constants import THROUGHPUT_GAIN
+from model_analyzer.perf_analyzer.perf_config import PerfAnalyzerConfig
 
 
 class PerfAnalyzerConfigGenerator(ConfigGeneratorInterface):
-    """ Given Perf Analyzer configuration options, generates Perf Analyzer configs """
+    """ 
+    Given Perf Analyzer configuration options, generates Perf Analyzer configs 
+    
+    All combinations are created at time 0, but it may return is_done==true
```
Comment (nv-braf) [inline model_analyzer/config/generate/perf_analyzer_config_generator.py:27]: change time 0 to `__init__`

File: model_analyzer/config/generate/perf_analyzer_config_generator.py [hunk 2]
```diff
@@ -66,7 +81,8 @@ def set_last_results(self, measurement):
         ----------
         measurement: Measurement from the last run
         """
-        pass
+        self._last_results = measurement
+        self._all_results.extend(measurement)
```
Comment (nv-braf) [inline model_analyzer/config/generate/perf_analyzer_config_generator.py:85]: Why not use `.append`? Isn't measurement just a single value?

File: model_analyzer/config/generate/perf_analyzer_config_generator.py [hunk 3]
```diff
@@ -114,3 +130,27 @@ def _create_perf_config_params(self):
...
         return perf_config_params
+
+    def _all_configs_returned(self):
+        return len(self._configs) == self._curr_config_index
+
+    def _last_results_erroneous(self):
+        return not self._last_results
+
+    def _throughput_gain_valid(self):
+        if len(self._all_results) < 4:
```
Comment (nv-braf) [inline model_analyzer/config/generate/perf_analyzer_config_generator.py:141]: Create a define for this

File: model_analyzer/config/generate/perf_analyzer_config_generator.py [hunk 4]
```diff
@@ -114,3 +130,27 @@ def _create_perf_config_params(self):
...
+            return True
+
+        return self._calculate_throughput_gain(1) > THROUGHPUT_GAIN or \
```
Comment (nv-braf) [inline model_analyzer/config/generate/perf_analyzer_config_generator.py:144]: Create a list comprehension to do this

File: model_analyzer/config/generate/perf_analyzer_config_generator.py [hunk 5]
```diff
@@ -114,3 +130,27 @@ def _create_perf_config_params(self):
...
+            self._calculate_throughput_gain(2) > THROUGHPUT_GAIN or \
+            self._calculate_throughput_gain(3) > THROUGHPUT_GAIN
+
+    def _calculate_throughput_gain(self, index):
```
Comment (nv-braf) [inline model_analyzer/config/generate/perf_analyzer_config_generator.py:148]: This is hard to understand - add a comment and rename member variable

File: model_analyzer/config/generate/perf_analyzer_config_generator.py (hunk 2 above)
Comment (tgerdesnv) [reply model_analyzer/config/generate/perf_analyzer_config_generator.py:85]: as we discussed, it is a list. I have updated the docstrings and variable name

**Pull request closed without merging**